### ISSUE TITLE
boto ConnectionContextWithHostAndPortBase seems busted

### ISSUE THREAD

**Issue by willkg** [mozilla/socorro-collector#17]
When setting up a self-contained docker environment that uses minio, we want to point the BotoS3CrashStorage at the minio container. Best way to do that is to use `ConnectionContextWithHostAndPortBase`.

First problem is that it extends `RequiredConfig` and thus has no functionality and doesn't support the API correctly. I think it should extend `ConnectionContextBase`.

Fixing that is cool, but then the logs quietly suggest there's a missing `_connect_to_endpoint` with no traceback:

```
crashmover_1  | 2016-06-10 00:25:12,223 ERROR - crashmover -  - Thread-2 - <class 'collector.external.boto.crashstorage.BotoS3CrashStorage'> failure: 'ConnectionContextWithHostAndPortBase' object has no attribute '_connect_to_endpoint'
```

This needs to get fixed for dockerizing to work.

Relatedly, I'm not wild about the quiet errors. I'd rather they were loud and had tracebacks. Definitely worth tracking down where they're erroring out and seeing what's going on in the logging code.

**Comment by willkg** [comment]
https://github.com/minio/minio#how-to-use-aws-cli-with-minio

https://github.com/minio/minio/issues/1025

**Comment by willkg** [comment]
Ugh this is super painful.

Things to know:

1. We're using boto 2.20 and they're up to 2.40 now.
2. The existing `ConnectionContextWithHostAndPortBase` isn't complete, so I nixed it.
3. I wrote a new `HostPortS3ConnectionContext` with as few bits as possible. That "works".
4. minio requires v4 of the aws auth api, so you need `S3_USE_SIGV4=True` in the environment.
5. minio doesn't have subdomain calling format (e.g. 'foo' bucket is not 'foo.host'), so need to use `OrdinaryCallingFormat` otherwise there is no host for the docker container you want.
6. Because we're specifying our own host and v4, it needs to stick the region in the auth something or other, but can't because we're not using a region. Only way I could deal with this is monkeypatching. YAY! There might be another way, but I wasn't able to get anything else to work.

That mostly works. Now I'm at the point where the minio docker container sometimes works and sometimes doesn't for reasons I don't understand. Further, when it doesn't work, the minio code goes to log the error, but there's a bug in the logging and so then it is too sad about the bug in the logging to tell me about the original error, so I have no idea what the problem is. Someone submitted a fix for that. Hopefully they update the docker container soon.

**Comment by willkg** [comment]
I should try fake-s3: https://hub.docker.com/r/lphoward/fake-s3/

**Comment by willkg** [comment]
Got it working with fake-s3. I'll put this with the docker branch and land the whole thing all at once.

**Comment by willkg** [comment]
Landed in #19.
